### PR TITLE
Implement skip graph

### PR DESCRIPTION
**Pull request by pipermerriam**
## What was wrong?

We need to implement a [Skip Graph](https://www.cs.yale.edu/homes/aspnes/papers/skip-graphs-journal.pdf) in order to support range queries over the DHT.

## How was it fixed?

Figured it out.

#### Cute Animal Picture

funny-cats-dogs-stuck-furniture-32

### CODE REVIEW

File: alexandria/skip_graph.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+            # else on the next level.  This suggests we need to avoid
+            # populating the `null` levels and only have firm links.
+            next_left_key = left.get_left_neighbor(level)
+            if next_left_key is None:
+                left = None
+                break
+            else:
+                left = self.nodes[next_left_key]
+
+        while right is not None and right.get_membership_at_level(level + 1) != node_membership:
```
Comment (lithp) [inline alexandria/skip_graph.py:180]: This is a small optimization, but once we know our predecessor at `level + 1`, we don't have to scan through nodes until we find the correct right. The correct right will be `left.get_right_neighbor(level + 1)`!

File: alexandria/skip_graph.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+            if left is not None or right is not None:
+                self._link_nodes(left, right, level)
+
+        self.nodes.pop(key)
+
+    def search(self, key: Key, current: SGNode) -> None:
+        self.logger.debug("Searching: %d", key)
+        return self._search(key, current, current.max_level)
+
+    def _search(self, key: Key, current: SGNode, level: int) -> SGNode:
```
Comment (lithp) [inline alexandria/skip_graph.py:279]: I'm not sure exactly how to do it, but you might be able to unify this with `_search_insert_point`, these two methods have a lot in common!

Maybe `_find_node` could return `left_neighbor, current, right_neighbor`. `_search_insert_point` interprets that result to either throw `AlreadyPresent` or return the neighbors. `_search` throws away the neighbors and throws `NotFound` or returns the node.

File: alexandria/skip_graph.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+    def delete(self, key: Key, current: SGNode) -> None:
+        self.logger.debug('Deleting: %d', key)
+        node = self.search(key, current)
+
+        left: Optional[SGNode]
+        right: Optional[SGNode]
+
+        for level in range(node.max_level, -1, -1):
+            left_neighbor_key = node.get_left_neighbor(level)
```
Comment (lithp) [inline alexandria/skip_graph.py:258]: This pattern shows up pretty often! A lot of these methods would get shorter if there was a method like `Graph.left_neighbor(self, node, level) -> Optional[SGNode]`.

File: alexandria/skip_graph.py (hunk 1 above)
Comment (pipermerriam) [reply alexandria/skip_graph.py:180]: fancy.  nice catch/optimization.

File: alexandria/skip_graph.py (hunk 3 above)
Comment (pipermerriam) [reply alexandria/skip_graph.py:258]: Yup, this made for easier to read code.

File: alexandria/skip_graph.py (hunk 2 above)
Comment (pipermerriam) [reply alexandria/skip_graph.py:279]: I'm now exposing this as a top level API `GraphAPI.find`.  It'll be a necessary external API for doing range queries.

**Pull request merged**